### PR TITLE
SOLR-15031: Prevent null being wrapped in a QueryValueSource

### PR DESCRIPTION
**Pull request by vonbox**
# Description

When parsing a sub query in a function query, FunctionQParser#parseValueSource does not check if the produced query object is null. When it is, it just wraps a null in a QueryValueSource object. This is a cause for NPE's in code consuming that object. Parsed queries can be null, for example when the query string only contains stopwords, so we need handle that condition.

# Solution

Added a null check to FunctionQParser#parseValueSource

# Tests

The problem can be reproduced via:
1. Start solr with the techproducts example collection: solr start -e techproducts
2. Add a stopword to SOLR_DIR/example/techproducts/solr/techproducts/conf/stopwords.txt, for example "at"
3. Execute a function query: http://localhost:8983/solr/techproducts/select?fieldquery={!field%20f=features%20v=%27%22at%22%27}&q={!func}%20if($fieldquery,1,0)

After applying this PR the NPE is gone.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java [hunk 1]
```diff
@@ -361,7 +361,9 @@ protected ValueSource parseValueSource(int flags) throws SyntaxError {
         ((FunctionQParser)subParser).setParseMultipleSources(true);
       }
       Query subQuery = subParser.getQuery();
-      if (subQuery instanceof FunctionQuery) {
+      if (subQuery == null) {
+        valueSource = new DoubleConstValueSource(0.0f);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/FunctionQParser.java:365]: Why a Double?

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java [hunk 2]
```diff
@@ -361,7 +361,9 @@ protected ValueSource parseValueSource(int flags) throws SyntaxError {
...
+      } else if (subQuery instanceof FunctionQuery) {
         valueSource = ((FunctionQuery) subQuery).getValueSource();
       } else {
         valueSource = new QueryValueSource(subQuery, 0.0f);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/FunctionQParser.java:369]: Should we add a test in QueryValueSource constructor to require non-null?

**Review by madrob — CHANGES_REQUESTED** [review]
Thanks for this fix, can you add your reproduction from the JIRA issue as a unit test?

**Comment by madrob** [comment]
Overall the fix is definitely good, and I think it's correct, just a few minor questions about it for completeness. Thank you for opening the PR!

**Comment by vonbox** [comment]
> Thanks for this fix, can you add your reproduction from the JIRA issue as a unit test?

Thanks for looking into it. I surely agree that every fix should have a unit test. However, though I know how to get around in the Solr code, the test suite is unknown territory. Can you give me a hint where I can add such test (and maybe find some inspiration from existing tests)?

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java (hunk 1 above)
Comment (vonbox) [reply solr/core/src/java/org/apache/solr/search/FunctionQParser.java:365]: Having a second look I see that a ConstValueSource is a better alternative, I'll update this.

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java (hunk 2 above)
Comment (vonbox) [reply solr/core/src/java/org/apache/solr/search/FunctionQParser.java:369]: I think it would be good to catch problems at the gate, adding this extra check. However, this could result in a risk that these exceptions pop up in places that were (seemed) executing fine. Very small risk, if a QueryValueSource with a null query is created a NPE is inevitable, so I cannot imagine adding such check will cause new problems. Still, I did not dare to add it yet. Can we get a veto of someone on this? (In the meantime I'll try it and run the test suite.)

Comment (vonbox) [reply solr/core/src/java/org/apache/solr/search/FunctionQParser.java:369]: I pushed a commit adding the null check to the QueryValueSource constructor

**Comment by madrob** [comment]
> Can you give me a hint where I can add such test (and maybe find some inspiration from existing tests)?

You can add a test to https://github.com/apache/lucene-solr/blob/master/solr/core/src/test/org/apache/solr/search/function/TestFunctionQuery.java the `text` field in the schema for that test is already configured to use stopwords: `stopworda` and `stopwordb`

**Comment by vonbox** [comment]
Thanks for the hint. I added a test.

**Comment by madrob** [comment]
Thanks for adding the test! it passes for me when run alone, but not as part of the full class, can you verify in your environment as well?

`./gradlew :solr:core:test --tests TestFunctionQuery`

**Comment by madrob** [comment]
Also, would you mind adding an entry to `solr/CHANGES.txt` with how you would like proper credit/attribution? I'll make sure that it gets to the right section, so don't stress about that if you are unsure.

**Comment by vonbox** [comment]
> Thanks for adding the test! it passes for me when run alone, but not as part of the full class, can you verify in your environment as well?
> 
> `./gradlew :solr:core:test --tests TestFunctionQuery`

That's strange, sometimes it succeeds, but often it fails. It looks like documents from previous tests pop up in the result, which makes me wonder what the clearIndex call actually does? I took the easy way out by removing the numFound check. So now it is just making sure no exception is thrown when executing the function, which is a good enough test I think.
